src/displayBackend/common/PgDirSharedBuffer.cpp

```diff
@@ -0,0 +1,114 @@
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU General Public License for more details.
+ *
+ *   You should have received a copy of the GNU General Public License
+ *   along with this program; if not, write to the Free Software
+ *   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307 USA
+ *
+ * Copyright (C) 2020 EPAM Systems Inc.
+ */
+
+#include "PgDirSharedBuffer.hpp"
+
+#include <iomanip>
+#include <vector>
+
+#include <xen/be/Exception.hpp>
+
+#include "DisplayItf.hpp"
+
+using std::min;
+
+using XenBackend::XenGnttabBuffer;
+
+void pgDirGetBufferRefs(domid_t domId, grant_ref_t startDirectory,
+						uint32_t size, GrantRefs& refs)
+{
+	XenBackend::Log log("PgDirSharedBuffer");
+
+	refs.clear();
+
+	size_t requestedNumGrefs = (size + XC_PAGE_SIZE - 1) / XC_PAGE_SIZE;
+
+	DLOG(log, DEBUG) << "Get buffer refs, directory: " << startDirectory
+					  << ", size: " << size
+					  << ", in grefs: " << requestedNumGrefs;
+
+	while(startDirectory != 0 && requestedNumGrefs)
+	{
+		DLOG(log, DEBUG) << "startDirectory: " << startDirectory;
+
+		XenGnttabBuffer pageBuffer(domId, startDirectory);
+
+		xendispl_page_directory* pageDirectory =
+				static_cast<xendispl_page_directory*>(pageBuffer.get());
+
+		size_t numGrefs = min(requestedNumGrefs, (XC_PAGE_SIZE -
+							  offsetof(xendispl_page_directory, gref)) /
+							  sizeof(uint32_t));
+
+		DLOG(log, DEBUG) << "Gref address: " << pageDirectory->gref
+						  << ", numGrefs " << numGrefs;
+
+		refs.insert(refs.end(), pageDirectory->gref,
+					pageDirectory->gref + numGrefs);
+
+		requestedNumGrefs -= numGrefs;
+
+		startDirectory = pageDirectory->gref_dir_next_page;
+	}
+
+	DLOG(log, DEBUG) << "Get buffer refs, num refs: " << refs.size();
+}
+
+void pgDirSetBufferRefs(domid_t domId, grant_ref_t startDirectory,
+						uint32_t size, GrantRefs& refs)
+{
+	XenBackend::Log log("PgDirSharedBuffer");
+
+	size_t requestedNumGrefs = (size + XC_PAGE_SIZE - 1) / XC_PAGE_SIZE;
+
+	DLOG(log, DEBUG) << "Set buffer refs, directory: " << startDirectory
+					  << ", size: " << size
+					  << ", in grefs: " << requestedNumGrefs;
+
+	grant_ref_t *grefs = refs.data();
+
+	while(startDirectory != 0 && requestedNumGrefs)
+	{
+		DLOG(log, DEBUG) << "startDirectory: " << startDirectory;
+
+		XenGnttabBuffer pageBuffer(domId, startDirectory);
+
+		xendispl_page_directory* pageDirectory =
+				static_cast<xendispl_page_directory*>(pageBuffer.get());
+
+		size_t numGrefs = min(requestedNumGrefs, (XC_PAGE_SIZE -
+							  offsetof(xendispl_page_directory, gref)) /
+							  sizeof(uint32_t));
+
+		DLOG(log, DEBUG) << "Gref address: " << pageDirectory->gref
+						  << ", numGrefs " << numGrefs;
+
+		memcpy(pageDirectory->gref, grefs, numGrefs * sizeof(grant_ref_t));
+
+		requestedNumGrefs -= numGrefs;
```

Comment:
what is if requestedNumGref =1; and that requestedNumGref -= 100; ?

Reply:
This is a pure code move, no changes to behavior. Please file a bug